File: cpp/problem019.cpp
/*
    Copyright (c) 2015 - buyuk.dev@gmail.com
    ----------------------------------------
    How many sundays happend on first day of a month
    between 1 Jan 1901 and 31 Dec 2000?
*/


#include <iostream>
#include <vector>
#include <string>
#include <list>
#include <map>
#include <algorithm>
#include <cmath>
#include <fstream>
#include <sstream>



#define MAX(a,b) (a<b?b:a)
#define MIN(a,b) (a<b?a:b)



using namespace std;


const int months[12] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
inline bool isLeapYear(const int year) {
    return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
}

int monthLength(int year, int month) {
    int days = 0;
    if (month != 1) {
        days = months[month];
    }
    else if (isLeapYear(year)) {
        days = 29;
    }
    return days;
}

int main(int argc, char* argv[]) 
{
    ios_base::sync_with_stdio(0);

    int firstDay = 0, sundaysCounter = 0;
    for (int year=1900; year<2001; ++year) {
        for (int month=0; month<12; ++month) {
            firstDay += monthLength(year, month);
            firstDay %= 7;
            if (year > 1900 && firstDay == 6) {
                sundaysCounter ++;
            }
        }
    }
    
    cout << "counted months starting at sunday between 1901 and 2001: " << sundaysCounter << endl;
    return 0;
}

File: python/problem019.py
#
# Copyright (c) 2016 - buyuk.dev@gmail.com
# ----------------------------------------
# Counting Sundays
#

months = [ 
    'jan', 'feb', 'mar', 'apr', 'may', 'jun',
    'jul', 'aug', 'sep', 'oct', 'nov', 'dec' 
]

monthsLengths = { 
    'jan': 31, 'feb': 28, 'mar': 31, 'apr': 30, 
    'may': 31, 'jun': 30, 'jul': 31, 'aug': 31,
    'sep': 30, 'oct': 31, 'nov': 30, 'dec': 31 
}

daysOfWeek = { 
    'mon': 0, 'tue': 1, 'wed': 2, 'thu': 3, 
    'fri': 4, 'sat': 5, 'sun': 6 
}



def isLeapYear(year):

    return year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)


def getMonthLength(year, month):

    month = months[month]
    days = monthsLengths[month]

    if month == 'feb' and isLeapYear(year):
        days += 1

    return days


def countDays(begYear, endYear, day):

    first = sundays = 0
    
    for year in range(begYear, endYear):
        for month in range( len(months) ):

            first += getMonthLength(year, month)
            first = first % len(daysOfWeek)
            
            if year > 1900 and first == daysOfWeek[day]:
                
                sundays += 1

    return sundays


if __name__ == '__main__':

    sundays = countDays(1900, 2001, 'sun')

    print "1st day of a month was Sunday exactly ", sundays, " times."
